src/ripple/overlay/Slot.h

```diff
@@ -0,0 +1,635 @@
+    typename clock_type::time_point
+        lastSelected_;               // last time peers were selected,
+                                     // used to age the slot
+    SlotState state_;                // slot's state
+    SquelchHandler const& handler_;  // squelch/unsquelch handler
+};
+
+template <typename clock_type>
+void
+Slot<clock_type>::checkIdle(PublicKey const& validator)
+{
+    auto now = clock_type::now();
+    for (auto it = peers_.begin(); it != peers_.end();)
+    {
+        auto& peer = it->second;
+        auto id = it->first;
+        ++it;
+        if (now - peer.lastMessage > IDLED)
+            deletePeer(validator, id, false);
+    }
+}
+
+template <typename clock_type>
+void
+Slot<clock_type>::update(
+    PublicKey const& validator,
+    id_t id,
+    std::weak_ptr<Peer> const& peerPtr,
+    protocol::MessageType type)
+{
+    auto now = clock_type::now();
+    auto it = peers_.find(id);
+    // First message from this peer
+    if (it == peers_.end())
+    {
+        peers_.emplace(std::make_pair(
+            id, PeerInfo{peerPtr, PeerState::Counting, 0, now, now}));
+        initCounting();
+        return;
+    }
+    // Message from a peer with expired squelch
+    else if (
+        it->second.state == PeerState::Squelched && now > it->second.expire)
+    {
+        it->second.state = PeerState::Counting;
+        it->second.lastMessage = now;
+        initCounting();
+        return;
+    }
+
+    auto& peer = it->second;
+    peer.lastMessage = now;
+
+    if (state_ != SlotState::Counting || peer.state == PeerState::Squelched)
+        return;
+
+    if (++peer.count > MIN_MESSAGE_THRESHOLD)
+        considered_.insert(id);
+    if (peer.count == (MAX_MESSAGE_THRESHOLD + 1))
+        ++reachedThreshold_;
+
+    if (now - lastSelected_ > 2 * MAX_UNSQUELCH_EXPIRE)
+    {
+        initCounting();
+        return;
+    }
+
+    if (reachedThreshold_ == MAX_SELECTED_PEERS)
+    {
+        // Randomly select MAX_SELECTED_PEERS peers from considered.
+        // Exclude peers that have been idling > IDLED -
+        // it's possible that checkIdle() has not been called yet.
+        // If number of remaining peers != MAX_SELECTED_PEERS
+        // then reset the Counting state and let checkIdle() handle
+        // idled peers.
+        std::unordered_set<id_t> selected;
+        while (selected.size() != MAX_SELECTED_PEERS && considered_.size() != 0)
+        {
+            auto i =
+                considered_.size() == 1 ? 0 : rand_int(considered_.size() - 1);
+            auto it = std::next(considered_.begin(), i);
+            auto id = *it;
+            considered_.erase(it);
+            if (peers_.find(id) == peers_.end())
+            {
+                // TBD have to log this
+                continue;
+            }
+            // TBD : should we be stricter and include < x% of IDLED? (50%?)
+            if (now - peers_[id].lastMessage < IDLED)
```

Comment:
We already searched for `id` above. Instead of looking up again save an iterator above and use the iterator.